### PR TITLE
First pass at using platform abstraction for plugins

### LINKED ISSUE #92279: Add platform interface to plugin template
Currently the plugin template uses platform channels as the cross-platform API, including web. We should change it to use a platform interface internally, just as we would for a federated plugin; there's no reason that platform interfaces need federation, and adding it would:
- make it clearer that platform channels don't have to be the cross-platform interface
- demonstrate a better pattern for web
- make it clearer how to, e.g., use FFI for one implementation
- be more consistent with the eventual federated template, so the two templates are more similar.

/cc @ditman

### PR DESCRIPTION
**Pull request by fuzzybinary**
Change the flutter create --template plugin command to create a PlatformInterface as an abstraction for plugins.

Work to fix #92279

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> This doesn't hit some of the requests of the ticket including showing an FFI implementation

I think you misread the issue slightly. I wasn't saying that the template should demonstrate using FFI, but that using a platform interface in the template makes it easier for people developing plugins to adopt non-platform-channel implementations (such as FFI) on some platforms.

**Comment by Hixie** [comment]
This kind of change should be tested (even though similar code currently lacks tests).

**Comment by fuzzybinary** [comment]
I've added a test to check that we create the plugin platform interface file, and added lines that check that the generated projects and examples pass analysis and unit tests. Since the internal test uses the platform channel, I *think* this covers the functionality I'm looking to test.

**Comment by christopherfujino** [comment]
@stuartmorgan can you take another look at this one?

**Comment by christopherfujino** [comment]
@fuzzybinary it looks like you have merge conflicts, can you resolve them by rebasing upstream?

**Comment by stuartmorgan** [comment]
This is near the top of my review queue BTW; sorry it had taken so long to get back to.

**Comment by skia-gold** [comment]
Gold has detected about 59 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/92672

**Comment by fuzzybinary** [comment]
Rebased but it looks like some tests are failing on Mac. I'll come back and take a look Monday when I have some time.

**Comment by fuzzybinary** [comment]
Okay! Re-rebased and looks like tests are passing (no changes) so must have been an infra thing. Sorry for the delay in checking.

File: packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  /// class that extends [{{pluginDartClass}}Platform] when they register themselves.
+  static set instance({{pluginDartClass}}Platform instance) {
+    PlatformInterface.verifyToken(instance, _token);
+    _instance = instance;
+  }
+
+  Future<String?> get platformVersion {
+    throw UnimplementedError('platformVersion() has not been implemented.');
+  }
+}
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl:29]: Please add the missing final newlines to all of these files.

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:4]: You shouldn't need `meta` here; `flutter/foundation.dart` vends common annotations.

(I'm not sure how you were able to compile this template, since `meta` isn't in the pubspec.)

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import '{{projectName}}_platform_interface.dart';
+
+/// An implementation of [{{pluginDartClass}}Platform] that uses method channels.
+class MethodChannel{{pluginDartClass}} extends {{pluginDartClass}}Platform {
+  /// The method channel used to interact with the native platform.
+  @visibleForTesting
+  final methodChannel = const MethodChannel('{{projectName}}');
+
+  @override
+  Future<String?> get platformVersion async {
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:15]: Same comment here about getters; use `getPlatformVersion() async`.

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import 'dart:async';
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:1]: Is this actually being used?

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+/// An implementation of [{{pluginDartClass}}Platform] that uses method channels.
+class MethodChannel{{pluginDartClass}} extends {{pluginDartClass}}Platform {
+  /// The method channel used to interact with the native platform.
+  @visibleForTesting
+  final methodChannel = const MethodChannel('{{projectName}}');
+
+  @override
+  Future<String?> get platformVersion async {
+    final version = await methodChannel.invokeMethod('getPlatformVersion');
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:16]: Let's update this to `invokeMethod<String>`

File: packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl [hunk 6]
```diff
@@ -1,23 +1,27 @@
...
+  @override
+  Future<String?> get platformVersion => Future.value('42');
+}
 
+void main() {
+  late {{pluginDartClass}} {{pluginClassLowerCamelCase}};
+  late Mock{{pluginDartClass}}Platform fakePlatform;
+  
   setUp(() {
-    channel.setMockMethodCallHandler((MethodCall methodCall) async {
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl:11]: All of the code you've removed from this file should still exist, just as a file called `projectName_method_channel_test.dart` (written against the platform interface though).

File: packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl [hunk 7]
```diff
@@ -4,16 +4,25 @@
...
-
 import 'dart:async';
 
-import 'package:flutter/services.dart';
+import '{{projectName}}_platform_interface.dart';
 
 class {{pluginDartClass}} {
-  static const MethodChannel _channel = MethodChannel('{{projectName}}');
+  factory {{pluginDartClass}}() {
+    _singleton ??= {{pluginDartClass}}._();
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:13]: Why have you restructured this class? You could have kept exactly the same structure and just had the implementation call `{{pluginDartClass}}Platform.instance.platformVersion`, so this doesn't seem related to the goal of the PR.

File: packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl [hunk 8]
```diff
@@ -4,16 +4,25 @@
...
+  static {{pluginDartClass}}? _singleton;
+
+  static {{pluginDartClass}}Platform get _platform {
+    return {{pluginDartClass}}Platform.instance;
+  }
 
-  static Future<String?> get platformVersion async {
-    final String? version = await _channel.invokeMethod('getPlatformVersion');
-    return version;
+  Future<String?> get platformVersion {
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:25]: Let's change this from `get platformVersion` to `getPlatformVersion()`. Using a getter here is [against Flutter style](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#getters-feel-faster-than-methods) (not that we have to follow Flutter style in a package, but the arguments that are made there are generally applicable), and even more importantly it doesn't scale to anything with arguments, making it harder to follow as a modifiable starting point.

File: dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart [hunk 9]
```diff
@@ -70,7 +70,7 @@ class PluginPlatformInterfaceMacOS {
       );
       pluginImplPubspecContent = pluginImplPubspecContent.replaceFirst(
           '    platforms:\n',
-          '    implements: plugin_platform_interface\n'
+          '    implements: plugin_platform_base\n'
```
Comment (stuartmorgan) [inline dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart:73]: This is a really confusing name. The better fix here would be to change `plugin` to `aplugin` or `someplugin` everywhere in this test, and keep the standard suffixes.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Sorry again for the delay. Overall this looks great, but some specific comments.

@ditman should look at the web template changes.

File: packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl (hunk 7 above)
Comment (fuzzybinary) [reply packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:13]: I think I was just bringing it in line with other examples I'd seen for this interface. I'll back it out and it can be a separate PR if desired.

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl (hunk 4 above)
Comment (fuzzybinary) [reply packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:1]: Don't you need it for Future?

**Review by ditman — COMMENTED** [review]
The web implementation looks correct. This should be easier to demonstrate by creating a `plugin` with this template and ensuring that the example app works correctly with `flutter run -d chrome` 

Also: do we want to update the `plugin` template, or create a new template for this example, @stuartmorgan ?

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl (hunk 2 above)
Comment (fuzzybinary) [reply packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:4]: Not sure why it compiles, but meta is needed for @visibleForTesting.

Comment (stuartmorgan) [reply packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:4]: No it isn't; you can use `flutter/foundation.dart` as I said in my comment.

https://github.com/flutter/flutter/blob/master/packages/flutter/lib/foundation.dart#L20

File: dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart [hunk 10]
```diff
@@ -31,46 +31,46 @@ TaskFunction dartPluginRegistryTest({
             'io.flutter.devicelab',
             '--platforms',
             'macos',
-            'plugin_platform_implementation',
+            'aplugin_platform_implementation',
```
Comment (fuzzybinary) [inline dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart:34]: @stuartmorgan I think this is what you meant when you said use 'aplugin', can you confirm?

File: packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl (hunk 4 above)
Comment (stuartmorgan) [reply packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl]: Simple `Future` usage is vended by `dart:core` starting in Dart 2.1.

Comment (fuzzybinary) [reply packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl]: Nope, I'm wrong about that too. Removed.

**Comment by stuartmorgan** [comment]
> Also: do we want to update the `plugin` template, or create a new template for this example, @stuartmorgan ?

We should update the existing template. This makes it easier to support web, makes it easier to support Dart-only implementations (e.g., Linux or Windows FFI), makes hybrid Dart/native platform implementations easier, and provides a much simpler path to converting a plugin to federation later. The only downside is one unnecessary API layer to plumb things though in the case where someone doesn't want any of that, but that seems like a very small price to pay. And if someone is really sure they don't want it, they can easily rip it out.)

(Also, when we later add a federated template, having this structure in the non-federated version will mean they are more similar, which I think will be helpful in the ecosystem in general.)

File: dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart (hunk 10 above)
Comment (stuartmorgan) [reply dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart:34]: Yes, that's correct.

**Comment by stuartmorgan** [comment]
> This should be easier to demonstrate by creating a `plugin` with this template and ensuring that the example app works correctly with `flutter run -d chrome`

Do we have any e2e web tests currently? E.g., something like the Android and iOS tests that use https://github.com/flutter/flutter/blob/0e2f51dfd04eca2935318389a75e8bc112809342/dev/devicelab/lib/tasks/plugin_tests.dart ?

File: packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl [hunk 11]
```diff
@@ -1,19 +1,16 @@
 {{#no_platforms}}
-// You have generated a new plugin project without
-// specifying the `--platforms` flag. A plugin project supports no platforms is generated.
-// To add platforms, run `flutter create -t plugin --platforms <platforms> .` under the same
-// directory. You can also find a detailed instruction on how to add platforms in the `pubspec.yaml` at https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms.
+// You have generated a new plugin project without specifying the `--platforms`
+// flag. A plugin project supports no platforms is generated. To add platforms,
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:3]: This sentence doesn't actually make sense (a pre-existing issue); as long as you're adjusting the comment let's fix it. I think it was intended to say something like "A plugin project that supports no platform was generated."

File: packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+import '{{projectName}}_method_channel.dart';
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl:2]: Nit: Add a blank line above this one to separate package imports from local imports.

File: packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  static final Object _token = Object();
+
+  static {{pluginDartClass}}Platform _instance = MethodChannel{{pluginDartClass}}();
+
+  /// The default instance of [{{pluginDartClass}}Platform] to use.
+  ///
+  /// Defaults to [MethodChannel{{pluginDartClass}}].
+  static {{pluginDartClass}}Platform get instance => _instance;
+  
+  /// Platform-specific plugins should set this with their own platform-specific
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl:17]: Nit: s/plugins/implementations/

File: packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl [hunk 14]
```diff
@@ -1,43 +1,25 @@
...
-        );
-    }
+  static void registerWith(Registrar registrar) {
+    {{pluginDartClass}}Platform.instance = {{pluginDartClass}}Web();
   }
 
   /// Returns a [String] containing the version of the platform.
-  Future<String> getPlatformVersion() {
+  @override
+  Future<String?> getPlatformVersion() {
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:22]: @ditman Should we make this `async` so that it can just `return value;`, or is there a downside to that?

File: packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl [hunk 15]
```diff
@@ -1,23 +1,27 @@
...
+    with MockPlatformInterfaceMixin
+    implements {{pluginDartClass}}Platform {
 
-  TestWidgetsFlutterBinding.ensureInitialized();
+  @override
+  Future<String?> getPlatformVersion() => Future.value('42');
+}
 
+void main() {
+  late {{pluginDartClass}} {{pluginClassLowerCamelCase}};
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl:15]: There's no reason for the object under test to be part of the fixture, and it's a practice that tends to make tests harder to understand (and doesn't scale to testing non-trivially-constructed objects anway). You can just put:

```
{{pluginDartClass}} {{pluginClassLowerCamelCase}} = {{pluginDartClass}}();
```

at the start of each test case (in this case, all one of them).

File: packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl [hunk 16]
```diff
@@ -1,23 +1,27 @@
...
-      return '42';
-    });
-  });
-
-  tearDown(() {
-    channel.setMockMethodCallHandler(null);
+    fakePlatform = Mock{{pluginDartClass}}Platform();
+    {{pluginDartClass}}Platform.instance = fakePlatform;
+    {{pluginClassLowerCamelCase}} = {{pluginDartClass}}();
   });
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl:23]: We should add a test like this one:
https://github.com/flutter/plugins/blob/7686be7ecbe1cc049423715b5e2d897cb63a43c1/packages/file_selector/file_selector_platform_interface/test/file_selector_platform_interface_test.dart#L14-L16

as it's good to have that coverage to make sure the default behavior is as expected.

File: packages/flutter_tools/templates/template_manifest.json [hunk 17]
```diff
@@ -326,12 +326,16 @@
         "templates/plugin/ios.tmpl/.gitignore",
         "templates/plugin/ios.tmpl/Assets/.gitkeep",
         "templates/plugin/lib/projectName.dart.tmpl",
+        "templates/plugin/lib/projectName_platform_interface.dart.tmpl",
+        "templates/plugin/lib/projectName_method_channel.dart.tmpl",
+        "templates/plugin/LICENSE.tmpl",
```
Comment (stuartmorgan) [inline packages/flutter_tools/templates/template_manifest.json:331]: Were we just not creating this file then? Could you add a quick test for this fix as well?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
A few more things, but otherwise looks good to me.

File: packages/flutter_tools/templates/template_manifest.json (hunk 17 above)
Comment (fuzzybinary) [reply packages/flutter_tools/templates/template_manifest.json:331]: Looks like this might have been a mistake in the merge, as this file is now in templates/plugin_shared/LICENSE.

But, there wasn't a unit test for it either, so I added this file (and README.md) and removed this line.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM, thanks!

@ditman Can you do the second review here since you've already looked at parts of it?

**Comment by fuzzybinary** [comment]
@stuartmorgan I'm having trouble parsing some of the CI check failures. Do those all appear to be infrastructure issues or is there a failure I should be looking into?

**Comment by ditman** [comment]
```
CIPD('ensure_installed') failed with: failed to resolve flutter/android/sdk/all/linux-amd64@latest (line 1): no such ref
```

I've checked in CIPD [here](https://chrome-infra-packages.appspot.com/p/flutter/android/sdk/all/linux-amd64/+/) and it seems that the `latest` Ref is gone. I don't have access to the trace logs. This might have changed in the way upstream configures its tests (I know there has been changes in the last few days about this!)

```
Excellent news! A package dependency has been removed!
  cli_util
To make sure we do not accidentally add this dependency back in the future,
please remove this packages from the allow-list in dev/bots/allowlist.dart.
Thanks!
```

This seems to have been removed ~13 days ago, [here](https://github.com/flutter/flutter/commit/0eae276e68ad43236217fa0c032ef253af0d0d70).

@fuzzybinary I think this needs another merge from master. I can get it done tomorrow when I checkout the branch and give it another review. Thanks for the contribution (and apologies for the delay!)

File: packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl [hunk 18]
```diff
@@ -1,44 +1,26 @@
...
-  Future<dynamic> handleMethodCall(MethodCall call) async {
-    switch (call.method) {
-      case 'getPlatformVersion':
-        return getPlatformVersion();
-      default:
-        throw PlatformException(
-          code: 'Unimplemented',
-          details: '{{projectName}} for web doesn\'t implement \'${call.method}\'',
-        );
-    }
```
Comment (ditman) [inline packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:36]: Good riddance!

**Review by ditman — APPROVED** [review]
This LGTM! I've also used it to generate a plugin, and the output is very cool!

Screen Shot 2022-03-09 at 12 17 27 PM

**Comment by ditman** [comment]
> Do we have any e2e web tests currently? E.g., something like the Android and iOS tests that use https://github.com/flutter/flutter/blob/0e2f51dfd04eca2935318389a75e8bc112809342/dev/devicelab/lib/tasks/plugin_tests.dart ?

AFAIK: no we don't. But I can't remember when this template changed last (over a year ago?)
